awx/sso/backends.py

```diff
@@ -423,3 +468,52 @@ def on_populate_user(sender, **kwargs):
     if profile.ldap_dn != ldap_user.dn:
         profile.ldap_dn = ldap_user.dn
         profile.save()
+
+    reconcile_users_org_team_mappings(user, desired_org_states, desired_team_states, 'LDAP')
+
+
+def reconcile_users_org_team_mappings(user, desired_org_states, desired_team_states, source):
+    from awx.main.models import Organization, Team
+
+    org_content_type = ContentType.objects.get_for_model(Organization)
+    team_content_type = ContentType.objects.get_for_model(Team)
+
+    # users_roles is a flat list of IDs
+    users_roles = list(user.roles.filter(content_type__in=[org_content_type, team_content_type]).values_list('pk', flat=True))
+
+    for object_type, desired_states, model in [('organization', desired_org_states, Organization), ('team', desired_team_states, Team)]:
+        # Get all of the roles in the desired states for efficient DB extraction
+        roles = []
+        for sub_dict in desired_states.values():
+            for role_name in sub_dict:
+                if sub_dict[role_name] == None:
+                    continue
+                if role_name not in roles:
+                    roles.append(role_name)
+
+        # Get a set of named tuples for the org name plus all of the roles we got above
+        model_roles = getattr(model, 'objects').filter(name__in=desired_states.keys()).values_list(*(['name'] + roles), named=True)
```

Comment:
There's a lot of indirection going on here, so I'd prefer to clean it up a bit for readability. Even after someone gets past those readability issues, it uses Django feature that most people won't be familiar with.

So to reduce indirection, I'd suggest

```python
model_roles = model.objects.filter(name__in=desired_states.keys()).values_list('name', *roles, named=True)
```